File: src/main/java/ir/alireza/sadeghi/shiraz/AI/MonteCarlo.java
package ir.alireza.sadeghi.shiraz.ai;

import ir.alireza.sadeghi.shiraz.BoardMethods;
import ir.alireza.sadeghi.shiraz.GameMethods;
import ir.alireza.sadeghi.shiraz.Hexagon;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.ArrayList;
import java.util.Hashtable;
import java.util.Random;
/*
 * Performs a Monte-Carlo tree search.
 * Automatically cuts off after 100 consecutive moves, as to make it not too large (which would be computationally expensive).
 */

public class MonteCarlo {

	private static final Logger logger = LogManager.getLogger(MonteCarlo.class);

	//this tree stays the same during the whole game - although if it is really possible, it starts from scratch
	Tree<GameState> monteCarloTree;
	Random random = new Random();
	private final static double explorationParam = Math.sqrt(2);
	private boolean chooseForAIMove;
	
	//need to cutoff when the game takes too long - about hundred, to not make it too big
	private final int cutoff = 100;
	private int currentNode = 1;
	
	//construct the tree, using the initial board state - automatically happens in move
	public MonteCarlo(Node<GameState> root) {
		monteCarloTree = new Tree<GameState>(root);
	}
	
	//do the whole monte-carlo search-> for a set number of seconds-> stop expanding after this is reached
	public Node<GameState> MCTS (int numberSeconds) {
		//for a certain amount of time, this is performed
		chooseForAIMove = false;
		
		//will still finish current simulation
		double curr = System.currentTimeMillis();
		while ((System.currentTimeMillis() - curr) < (numberSeconds*1000)) {
			logger.trace("cutoff " + currentNode);
			currentNode = 1;
			selection(monteCarloTree.root);
		}
		
		chooseForAIMove = true;
		Node<GameState> chosen = sucChild(monteCarloTree.root);
		
		return chosen;
	}
	
	//start from the root
	//select successive child-nodes, until a lead node is reached
	//then, if it terminates the tree, it backpropagates or expands
	public void selection(Node<GameState> current){
		while (!current.isLeaf() && currentNode < cutoff) {
			currentNode++;
			current = sucChild(current);
		}
		if (!current.data.terminal && currentNode < cutoff) {
			currentNode++;
			simulation(current);
		}
	}
	
	//chooses successful child node - similar to the greedy algorithm
	public Node<GameState> sucChild(Node<GameState> parent){
		Node<GameState> bestChild = null;
		double maxEval = Double.NEGATIVE_INFINITY;
		ArrayList<Node<GameState>> multiple = new ArrayList<>();
		
		for (int i = 0; i < parent.children.size(); i++){
			if (Math.max(maxEval, parent.children.get(i).totValue) > maxEval && (!chooseForAIMove || maxEval < Double.POSITIVE_INFINITY )) {
				multiple.clear();
				maxEval = parent.children.get(i).totValue;
				bestChild = parent.children.get(i);
			} else if(Math.max(maxEval, parent.children.get(i).totValue) == maxEval && (!chooseForAIMove || maxEval < Double.POSITIVE_INFINITY )){
				multiple.add(parent.children.get(i));
			}
		}
		
		if(multiple.size()> 1) {
			bestChild = multiple.get(random.nextInt(multiple.size()));
		}
		
		return bestChild;
	}
	
	//expands the current node
	public void expansion(Node<GameState> expand) {
		AddNodes.addForOne(expand);
	}
	
	//simulation (also known as roll out)
	public void simulation(Node<GameState> current) {
		expansion(current);
		Node<GameState> newChoice = sucChild(current);
		
		if (newChoice.data.terminal || currentNode >= cutoff) {
			backpropagate(newChoice, newChoice.data.winner);
		}
		else {
			currentNode ++;
			simulation(newChoice);
		}
	}
	
	//backpropagation method - automatically calculate all the new values for it!!
	public void backpropagate(Node<GameState> toPropagate, int winningplayer) {
		if (!(toPropagate.parent == null)) {
			
			toPropagate.nVisits++;
			//if it wins, add the win as well
			if (winningplayer ==0) {
				toPropagate.wins =  (toPropagate.wins + 0.5);
			}
			if (GameMethods.changePlayer(toPropagate.returnData().turn) == winningplayer) {
				toPropagate.wins++;
			}
			
			toPropagate.calcMCTSvalue();
			backpropagate(toPropagate.parent, winningplayer);
		}
	}
	
	//wins = number of wins for the node considered after the i-th move
	//simulations = number of simulations for the node considered after the i-th move
	//parentSimulations = total number of simulations after the i-th move run by the parent node of the one considered
	public static double calculateUCB(double wins, int simulations, int parentSimulations) {
		return ((wins / simulations) + explorationParam * Math.sqrt(Math.log(parentSimulations)/simulations));
	}

	//change the current root node to the one that is needed
	public void changeRoot(Node<GameState> changed) {
		monteCarloTree.setRoot(changed);
		changed.parent = null;
	}
		
	//change the node from outside (so, the human player)
	public void changeRootOutside(Hashtable<String, Hexagon> board) {
		boolean isThere = false;
		int i = 0;
		while (!isThere && i < monteCarloTree.root.children.size()) {
			if (!monteCarloTree.root.isLeaf()) {
				if (BoardMethods.compareHashTables(monteCarloTree.getRoot().children.get(i).data.boardState, board)) {
					changeRoot(monteCarloTree.getRoot().children.get(i));
					isThere = true;
				}
			}
			i++;
		}
		if (!isThere) {
			GameState newRoot = new GameState(board,GameMethods.changePlayer(monteCarloTree.root.data.evaluateFrom));
			changeRoot(new Node<GameState>(newRoot));
		}
	}
}

File: src/main/java/ir/alireza/sadeghi/shiraz/AI/MonteCarlo.kt
package ir.alireza.sadeghi.shiraz.ai

import ir.alireza.sadeghi.shiraz.BoardMethods
import ir.alireza.sadeghi.shiraz.GameMethods
import ir.alireza.sadeghi.shiraz.Hexagon
import org.apache.logging.log4j.LogManager
import java.util.*
import kotlin.math.ln
import kotlin.math.sqrt

/*
 * Performs a Monte-Carlo tree search.
 * Automatically cuts off after 100 consecutive moves, as to make it not too large (which would be computationally expensive).
 */
class MonteCarlo(root: Node<GameState?>) {
    //this tree stays the same during the whole game - although if it is really possible, it starts from scratch
    private var monteCarloTree: Tree<GameState?>
    private var random = Random()
    private var chooseForAIMove = false

    //need to cutoff when the game takes too long - about a hundred, to not make it too big
    private val cutoff = 100
    private var currentNode = 1

    //construct the tree, using the initial board state - automatically happens in move
    init {
        monteCarloTree = Tree(root)
    }

    //do the whole monte-carlo search-> for a set number of seconds-> stop expanding after this is reached
    fun mcts(numberSeconds: Int): Node<GameState?>? {
        //for a certain amount of time, this is performed
        chooseForAIMove = false

        //will still finish current simulation
        val curr = System.currentTimeMillis().toDouble()
        while (System.currentTimeMillis() - curr < numberSeconds * 1000) {
            logger.trace("cutoff $currentNode")
            currentNode = 1
            selection(monteCarloTree.root)
        }
        chooseForAIMove = true
        return sucChild(monteCarloTree.root)
    }

    //start from the root
    //select successive child-nodes, until a lead node is reached
    //then, if it terminates the tree, it back propagates or expands
    private fun selection(current: Node<GameState?>?) {
        var c = current
        while (!c!!.isLeaf && currentNode < cutoff) {
            currentNode++
            c = sucChild(c)
        }
        if (!c.data!!.terminal && currentNode < cutoff) {
            currentNode++
            simulation(c)
        }
    }

    //chooses successful child node - similar to the greedy algorithm
    private fun sucChild(parent: Node<GameState?>?): Node<GameState?>? {
        var bestChild: Node<GameState?>? = null
        var maxEval = Double.NEGATIVE_INFINITY
        val multiple = ArrayList<Node<GameState?>?>()
        for (i in parent!!.children!!.indices) {
            if (maxEval.coerceAtLeast(parent.children!![i].totValue) > maxEval && (!chooseForAIMove || maxEval < Double.POSITIVE_INFINITY)
            ) {
                multiple.clear()
                maxEval = parent.children!![i].totValue
                bestChild = parent.children!![i]
            } else if (maxEval.coerceAtLeast(parent.children!![i].totValue) == maxEval && (!chooseForAIMove || maxEval < Double.POSITIVE_INFINITY)
            ) {
                multiple.add(parent.children!![i])
            }
        }
        if (multiple.size > 1) {
            bestChild = multiple[random.nextInt(multiple.size)]
        }
        return bestChild
    }

    //expands the current node
    private fun expansion(expand: Node<GameState?>?) {
        AddNodes.addForOne(expand)
    }

    //simulation (also known as roll out)
    private fun simulation(current: Node<GameState?>?) {
        expansion(current)
        val newChoice = sucChild(current)
        if (newChoice!!.data!!.terminal || currentNode >= cutoff) {
            backPropagate(newChoice, newChoice.data!!.winner)
        } else {
            currentNode++
            simulation(newChoice)
        }
    }

    //backpropagation method - automatically calculate all the new values for it!!
    private fun backPropagate(toPropagate: Node<GameState?>?, winningPlayer: Int) {
        if (toPropagate!!.parent != null) {
            toPropagate.nVisits++
            //if it wins, add the win as well
            if (winningPlayer == 0) {
                toPropagate.wins = toPropagate.wins + 0.5
            }
            if (GameMethods.changePlayer(toPropagate.returnData()!!.turn) == winningPlayer) {
                toPropagate.wins++
            }
            toPropagate.calcMCTSvalue()
            backPropagate(toPropagate.parent, winningPlayer)
        }
    }

    //change the current root node to the one that is needed
    private fun changeRoot(changed: Node<GameState?>?) {
        monteCarloTree.root =changed
        changed!!.parent = null
    }

    //change the node from outside (so, the human player)
    fun changeRootOutside(board: Hashtable<String?, Hexagon?>?) {
        var isThere = false
        var i = 0
        while (!isThere && i < monteCarloTree.root?.children!!.size) {
            if (!monteCarloTree.root!!.isLeaf) {
                if (BoardMethods.compareHashTables(monteCarloTree.root!!.children?.get(i)?.data?.boardState, board)) {
                    changeRoot(monteCarloTree.root!!.children?.get(i))
                    isThere = true
                }
            }
            i++
        }
        if (!isThere) {
            val newRoot = GameState(board, GameMethods.changePlayer(monteCarloTree.root?.data!!.evaluateFrom))
            changeRoot(Node(newRoot))
        }
    }

    companion object {
        private val logger = LogManager.getLogger(MonteCarlo::class.java)
        private val explorationParam = sqrt(2.0)

        //wins = number of wins for the node considered after the i-th move
        //simulations = number of simulations for the node considered after the i-th move
        //parentSimulations = total number of simulations after the i-th move run by the parent node of the one considered
        fun calculateUCB(wins: Double, simulations: Int, parentSimulations: Int): Double {
            return wins / simulations + explorationParam * sqrt(ln(parentSimulations.toDouble()) / simulations)
        }
    }
}